pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go

```diff
@@ -0,0 +1,234 @@
+				// so we should not get to this point.
+				// But if we do (e.g manual status update), we also delete the status from the ShootExtensionStatus resource
+				s.log.Debugf(fmt.Sprintf("Deleting the provider status for the %q extension from the ShootExtensionStatus for Shoot %q", kind, shoot.Name))
+				shootExtensionStatus.Statuses = removeExtensionStatus(shootExtensionStatus.Statuses, kind)
+				return s.patchShootExtensionStatus(ctx, shootExtensionStatus, shootExtensionStatusCopy, kind)
+			}
+		}
+
+		newEntry := gardencorev1alpha1.ExtensionStatus{
+			Kind:    kind,
+			Type:    extensionType,
+			Purpose: purpose,
+			Status:  *currentProviderStatus,
+		}
+
+		// new provider status that has not yet been synced
+		// add the new ExtensionStatus entry
+		if currentExtensionStatus == nil {
+			shootExtensionStatus.Statuses = append(shootExtensionStatus.Statuses, newEntry)
+			return s.patchShootExtensionStatus(ctx, shootExtensionStatus, shootExtensionStatusCopy, kind)
+		}
+
+		// at this point we know that we have a provider status that also exists in the ShootExtensionStatus in the Garden cluster
+		// we need to check if the status is up to date by checking for equality
+		if apiequality.Semantic.DeepEqual(*currentExtensionStatus, newEntry) {
+			s.log.Debugf("Skipping sync of the ShootExtensionStatus for the %q extension of Shoot %q. The resource is up-to-date", kind, shoot.Name)
+			return reconcile.Result{}, nil
+		}
+
+		// patch the existing ExtensionStatus entry and patch the ShootExtensionStatus resource
+		shootExtensionStatus.Statuses[indexOldStatus] = newEntry
+		return s.patchShootExtensionStatus(ctx, shootExtensionStatus, shootExtensionStatusCopy, kind)
+	}
+}
+
+// getShootForRequest returns the Shoot resource for a reconcile.Request parsed from the Cluster resource in the Seed
+func (s *ShootExtensionStatusControl) getShootForRequest(ctx context.Context, req reconcile.Request) (*gardencorev1beta1.Shoot, error) {
+	cluster, err := extensions.ClusterFromRequest(ctx, s.seedClient, req)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		return nil, fmt.Errorf("could not get cluster with name %s : %w", cluster.Name, err)
+	}
+
+	shoot, err := extensions.ShootFromCluster(s.decoder, cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	if shoot == nil {
+		return nil, fmt.Errorf("cluster resource %s doesn't contain shoot resource", cluster.Name)
+	}
+	return shoot, nil
+}
+
+func (s *ShootExtensionStatusControl) patchShootExtensionStatus(ctx context.Context, shootExtensionStatus *gardencorev1alpha1.ShootExtensionStatus, shootExtensionStatusCopy *gardencorev1alpha1.ShootExtensionStatus, kind string) (reconcile.Result, error) {
+	if err := s.gardenClient.Patch(ctx, shootExtensionStatus, client.MergeFromWithOptions(shootExtensionStatusCopy, client.MergeFromWithOptimisticLock{})); err != nil {
```

Comment:
This has a high chance of conflicting with other current updates to the `ShootExtensionStatus`.
Have you considered a strategic merge patch without optimistic locking?
This will need some `patchMergeKey` here: https://github.com/gardener/gardener/blob/abfb062f5121cf53b7e52ddf65022c3f6b867bb4/pkg/apis/core/v1alpha1/types_shootextensionstatus.go#L34

Reply:
Yeah, I tried using the `strategicMergePatch` but there is an issue with using `rawExtension` field. That is also the reason why in the ShootState a json-merge-patch is used @plkokanov .

It seems like when using `strategicMergePatch`  it also tries to merge the `rawExtension` field - and it does not know how to do that for lists inside the field (instead of simply replacing the field content).

This can be observed consequently only when there is a conflict with an existing list entry.

>  unable to find api field in struct RawExtension for the json field “ec2”

If you know how to use the `strategicMergePatch` safely with `rawExtension` fields, I would of course prefer that.

Reply:
Yeah, checked this in detail now and it seems to be very tricky indeed.

I see several problems:
- we would need a `UID` field in `ExtensionStatus` which can be used as a `patchMergeKey` for `ShootExtensionStatus.Statuses` (strategic merge can't handle multiple merge keys, I think)
- we would need to use the [`replace` directive](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/strategic-merge-patch.md#replace-directive) to tell strategic merge patch to not merge but replace the `RawExtension` field. The problem here is, that adding it to the `Status` field (`patchStrategy:"replace"`) is not sufficient, as it will send all statuses on every patch. The directive has to be added inside `RawExtension`'s json. However, then the patch computation will still fail, as the computation doesn't take the directives into account but only the patch strategy.

TL;DR: this is just to confusing to figure out, so let's use the merge patch with optimistic locking as you proposed 👍